File: backend/services/maritime_math.py
from typing import Dict, List, Any
import math

VESSEL_CLASSES = {
    "Capesize": {"capacity_dwt": 150000, "laden_draft_m": 18.0, "daily_cost_usd": 25000, "block_coeff": 0.85},
    "Panamax":  {"capacity_dwt": 75000,  "laden_draft_m": 14.0, "daily_cost_usd": 15000, "block_coeff": 0.82},
    "Supramax": {"capacity_dwt": 50000,  "laden_draft_m": 11.5, "daily_cost_usd": 12000, "block_coeff": 0.80},
    "Handysize":{"capacity_dwt": 30000,  "laden_draft_m": 10.0, "daily_cost_usd": 9000,  "block_coeff": 0.78}
}

def calculate_brackish_sinkage(draft_laden: float, port_density: float) -> float:
    """
    Calculates Fresh Water Allowance (FWA) / Brackish Water Sinkage.
    Higher sinkage occurs in riverine ports like Haldia (density ~1.005-1.015 g/cm3).
    """
    if port_density >= 1.025:
        return 0.0
    return draft_laden * ((1.025 - port_density) / port_density)


def calculate_hydrodynamic_squat(block_coefficient: float, speed_knots: float) -> float:
    """
    Calculates vessel squat in shallow, confined fairways.
    """
    return (2 * block_coefficient * (speed_knots ** 2)) / 100


def calculate_dynamic_ukc(charted_depth: float, tidal_height: float, draft_laden: float, delta_draft: float, squat: float) -> float:
    """
    Calculates Dynamic Under Keel Clearance (UKC).
    """
    arrival_draft = draft_laden + delta_draft + squat
    return charted_depth + tidal_height - arrival_draft


def evaluate_vessel_safety(
    charted_depth: float, 
    tidal_height: float, 
    draft_laden: float, 
    port_density: float, 
    block_coeff: float = 0.85, 
    speed_knots: float = 12.0, 
    min_ukc_margin: float = 1.0
) -> Dict[str, Any]:
    """
    Evaluates end-to-end hydrodynamic safety constraints.
    """
    delta_draft = calculate_brackish_sinkage(draft_laden, port_density)
    squat = calculate_hydrodynamic_squat(block_coeff, speed_knots)
    ukc_dynamic = calculate_dynamic_ukc(charted_depth, tidal_height, draft_laden, delta_draft, squat)
    arrival_draft = draft_laden + delta_draft + squat
    max_permissible_draft = charted_depth + tidal_height - min_ukc_margin
    
    return {
        "is_safe": ukc_dynamic >= min_ukc_margin,
        "arrival_draft_m": round(arrival_draft, 3),
        "delta_draft_m": round(delta_draft, 3),
        "squat_m": round(squat, 3),
        "dynamic_ukc_m": round(ukc_dynamic, 3),
        "max_permissible_draft_m": round(max_permissible_draft, 3),
        "required_margin_m": min_ukc_margin
    }


def calculate_cargo_split(
    total_volume_mt: float,
    dest_port_depth: float,
    tidal_height: float,
    port_density: float,
    speed_knots: float = 12.0
) -> Dict[str, Any]:
    """
    Evaluates fleet candidates and determines cargo splitting strategy when single-vessel fixtures fail.
    """
    feasible_vessels = []
    
    for class_name, specs in VESSEL_CLASSES.items():
        safety = evaluate_vessel_safety(
            charted_depth=dest_port_depth,
            tidal_height=tidal_height,
            draft_laden=specs["laden_draft_m"],
            port_density=port_density,
            block_coeff=specs["block_coeff"],
            speed_knots=speed_knots
        )
        if safety["is_safe"]:
            feasible_vessels.append((class_name, specs, safety))
            
    if not feasible_vessels:
        return {
            "feasible": False,
            "strategy": "Offshore Lighterage Required (Sandheads Transshipment)",
            "reason": "No bulk carrier class satisfies port depth and UKC safety constraints.",
            "recommended_vessels": []
        }
        
    # Pick the largest feasible vessel for maximum economies of scale
    feasible_vessels.sort(key=lambda x: x[1]["capacity_dwt"], reverse=True)
    best_class, best_specs, best_safety = feasible_vessels[0]
    
    vessel_count = math.ceil(total_volume_mt / best_specs["capacity_dwt"])
    estimated_daily_cost = vessel_count * best_specs["daily_cost_usd"]
    
    return {
        "feasible": True,
        "strategy": f"Direct Fixture ({best_class})" if vessel_count == 1 else f"Split Cargo into {vessel_count}x {best_class}",
        "primary_vessel_class": best_class,
        "vessel_count": vessel_count,
        "total_volume_mt": total_volume_mt,
        "calculated_arrival_draft_m": best_safety["arrival_draft_m"],
        "max_permissible_draft_m": best_safety["max_permissible_draft_m"],
        "clearance_margin_m": best_safety["dynamic_ukc_m"],
        "estimated_daily_cost_usd": estimated_daily_cost,
        "demurrage_risk": "Low" if best_safety["dynamic_ukc_m"] >= 1.5 else "Moderate"
    }

File: backend/services/maritimeMath.js
/**
 * Maritime Math & Constraint Solver Service
 * Implements physical constraints for bulk carriers in shallow fairways.
 */

const { PrismaClient } = require('@prisma/client');
const prisma = new PrismaClient();

let cachedFleet = null;
let lastFleetFetch = 0;

async function getFleet() {
    const now = Date.now();
    // Cache for 1 hour to prevent DB spam
    if (!cachedFleet || (now - lastFleetFetch) > 3600000) {
        cachedFleet = await prisma.vessel.findMany();
        lastFleetFetch = now;
    }
    return cachedFleet;
}

/**
 * Calculates Brackish Water Sinkage (Fresh Water Allowance)
 */
function calculateBrackishSinkage(draftLaden, portDensity) {
    // 1.025 is standard seawater density
    return draftLaden * ((1.025 - portDensity) / portDensity);
}

/**
 * Calculates Hydrodynamic Squat Effect for shallow fairways
 */
function calculateHydrodynamicSquat(blockCoeff, speedKnots) {
    return (2 * blockCoeff * Math.pow(speedKnots, 2)) / 100;
}

/**
 * Calculates Dynamic Under Keel Clearance (UKC)
 */
function calculateDynamicUKC(chartedDepth, tidalHeight, draftLaden, deltaDraft, squat) {
    return (chartedDepth + tidalHeight) - (draftLaden + deltaDraft + squat);
}

/**
 * Evaluates the fleet against destination port constraints
 */
async function evaluateRequisition(volume_mt, dest_port_draft, commodity, lat = 21.02, lon = 88.06, brackishDensity = 1.025, chartedDepth = 15.0) {
    const fleet = await getFleet();
    const UKC_MARGIN = 1.0;  // 1.0 meter safety margin
    const PORT_DENSITY = brackishDensity; // Using DB density instead of hardcoded
    
    // Fetch real-time wave/tide data from Open-Meteo Marine API
    let TIDAL_HEIGHT = 1.5; // fallback
    try {
        // Using wave_height as a proxy for dynamic tidal action at the port entrance
        const res = await fetch(`https://marine-api.open-meteo.com/v1/marine?latitude=${lat}&longitude=${lon}&hourly=wave_height`);
        if (res.ok) {
            const data = await res.json();
            if (data.hourly && data.hourly.wave_height && data.hourly.wave_height.length > 0) {
                TIDAL_HEIGHT = data.hourly.wave_height[0] || 1.5;
            }
        }
    } catch (e) {
        console.error("Failed to fetch live tide data, using fallback.", e.message);
    }
    let validVessels = [];

    for (const vessel of fleet) {
        // Calculate Brackish Water Sinkage using dynamic port density
        const deltaDraft = calculateBrackishSinkage(vessel.laden_draft, PORT_DENSITY);

        // Calculate Hydrodynamic Squat
        const squat = calculateHydrodynamicSquat(vessel.block_coeff, vessel.speed_knots);

        // Calculate Dynamic Under Keel Clearance using dynamic charted depth
        const ukcDynamic = calculateDynamicUKC(chartedDepth, TIDAL_HEIGHT, vessel.laden_draft, deltaDraft, squat);
        
        if (ukcDynamic >= UKC_MARGIN) {
            validVessels.push({
                ...vessel,
                calculatedDraft: Number((vessel.laden_draft + deltaDraft).toFixed(2)),
                clearance_margin: Number(ukcDynamic.toFixed(2))
            });
        }
    }

    if (validVessels.length === 0) {
        return {
            feasible: false,
            strategy: "Offshore Transshipment Required (e.g., Lighterage at Sandheads)",
            details: "No standard vessel class clears dynamic UKC limits for this port.",
            calculatedDraft: null
        };
    }

    // Sort by cost efficiency per metric ton
    validVessels.sort((a, b) => (a.daily_cost / a.capacity) - (b.daily_cost / b.capacity));
    
    const bestVessel = validVessels[0];
    const vesselCount = Math.ceil(volume_mt / bestVessel.capacity);

    let strategy = `Direct Fixture: 1x ${bestVessel.name}`;
    if (vesselCount > 1) {
        strategy = `Split Cargo into ${vesselCount}x ${bestVessel.name}s`;
    }

    return {
        feasible: true,
        strategy: strategy,
        vessel_class: bestVessel.name,
        total_vessels: vesselCount,
        calculatedDraft: bestVessel.calculatedDraft,
        clearance_margin: bestVessel.clearance_margin,
        portMaxDraft: dest_port_draft
    };
}

module.exports = {
    calculateBrackishSinkage,
    calculateHydrodynamicSquat,
    calculateDynamicUKC,
    evaluateRequisition
};
